File: javascript/phone-number/phone-number.js
function PhoneNumber (string) {
    'use strict';

    this.string = string;

    this.number = function () {
        return validDigits(integerCharacters(this.string));
    }

    this.areaCode = function () {
        return this.number().slice(0,3);
    }

    this.prefix_ = function () {
        return this.number().slice(3,6);
    }

    this.suffix = function () {
        return this.number().slice(6);
    }

    this.toString = function () {
        return '(' + this.areaCode() + ') ' +
                     this.prefix_() + '-' +
                     this.suffix();
    }

    function integerCharacters (string) {
        return string.replace(/[\-\s\(\)\.]/g,'');
    }

    function validDigits (string) {
        if (string.length === 10) {
            return string;
        } else if (string.length === 11 &&  string[0] === '1') {
            return string.slice(1);
        } else {
            return '0000000000';
        }
    }
}

module.exports = PhoneNumber;

File: ruby/phone-number/phone_number.rb
class PhoneNumber
  attr_reader :number

  def initialize num_str
    @number = process num_str
  end

  def area_code
    number[0..2]
  end

  def process num_str
    num_str.gsub! /[\.()\-\s]/, ''
    if num_str.length == 10 && !num_str[/[a-zA-Z]/]
      num_str
    elsif num_str.length == 11 && num_str[0] == '1'
      num_str[1..-1]
    else
      '0000000000'
    end
  end

  def to_s
    "(#{area_code}) #{number[3..5]}-#{number[6..9]}"
  end
end
